Use streaming IPC for log forwarding

pvollan opened the pull request:
#### 48c05b7a72999c1f72adbfe214f956a11c55ede3
<pre>
Use streaming IPC for log forwarding
<a href="https://rdar.apple.com/133129911">rdar://133129911</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=277592">https://bugs.webkit.org/show_bug.cgi?id=277592</a>

Reviewed by Chris Dumez.

Use streaming IPC for log forwarding to reduce power usage. Sending data over streaming IPC is more
efficient than sending using Mach messages.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::didFinishLoading):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setupLogStream):
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent): Deleted.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::setPreferenceValue):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::didReceiveResponse):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureNetworkProcessConnection):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/LaunchServicesDatabaseManager.mm:
(WebKit::LaunchServicesDatabaseManager::handleEvent):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerLogHook):
(WebKit::WebProcess::setupLogStream):
(WebKit::WebProcess::sendLogOnStream):
(WebKit::logQueue): Deleted.
(WebKit::registerLogHook): Deleted.

Canonical link: <a href="https://commits.webkit.org/284209@main">https://commits.webkit.org/284209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/17a4042d4fee7ddeb1798c3e73b529fc4a55860f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21343 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72753 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19829 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55872 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19645 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54755 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13183 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71751 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43928 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59302 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35219 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/68195 "Found 1 webkitpy test failure: webkit.messages_unittest.GeneratedFileContentsTest.test_message_argument_description") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40594 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18186 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62545 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17075 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74447 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12655 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16324 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62240 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-scale.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12695 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59383 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62267 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10219 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3834 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10480 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43877 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44951 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46145 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/LogStream.h:
```diff
@@ -0,0 +1,61 @@
+    RefPtr<IPC::StreamServerConnection> m_logStreamConnection;
+    RefPtr<IPC::StreamConnectionWorkQueue> m_logWorkQueue;
+    uint64_t m_pid { 0 };
+    std::optional<LogStreamIdentifier> m_logStreamIdentifier;
```
We could use `Markable<LogStreamIdentifier>`. It is more space efficient for data members.

pvollan replied:
Ah, good point! I will address this and the other comments in the next patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
```diff
@@ -217,6 +218,12 @@ template: enum class WebCore::ServiceWorkerRegistrationIdentifierType
 
+header: <wtf/ObjectIdentifier.h>
+template: enum class WebKit::LogStreamIdentifierType
+[WebKitPlatform, CustomHeader, AdditionalEncoder=StreamConnectionEncoder] class WTF::AtomicObjectIdentifier {
```
There is already AtomicObjectIdentifier in this file, why are you adding another one? Just add `template: enum class WebKit::LogStreamIdentifierType` to the existing one.

cdumez commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -8031,6 +8034,7 @@
 		E37A2F1A2B8523300087F394 /* NetworkingProcessExtension.entitlements */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text.plist.entitlements; path = NetworkingProcessExtension.entitlements; sourceTree = "<group>"; };
 		E37A2F1B2B8523B10087F394 /* WebContentProcessExtension.entitlements */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text.plist.entitlements; path = WebContentProcessExtension.entitlements; sourceTree = "<group>"; };
 		E37A2F1C2B8523B10087F394 /* GPUProcessExtension.entitlements */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text.plist.entitlements; path = GPUProcessExtension.entitlements; sourceTree = "<group>"; };
+		E38034DB2C63D6670095F1E0 /* LogStreamMessages.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; name = LogStreamMessages.h; path = ../../../../../../../../Build/Debug/DerivedSources/WebKit/LogStreamMessages.h; sourceTree = "<group>"; };
```
The `path` doesn't seem right here.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -832,49 +824,46 @@ static void registerLogHook()
-            msg.privdata_sz = privdata.size();
+        char* messageString = os_log_copy_message_string(msg);
+        if (messageString) {
+            std::span logString(reinterpret_cast<uint8_t*>(messageString), strlen(messageString) + 1);
```
reinterpret_cast -> byteCast

cdumez commented on Source/WebKit/NetworkProcess/LogStream.h:
```diff
@@ -0,0 +1,64 @@
+
+    void setup(uint64_t pid, IPC::StreamServerConnectionHandle&&, LogStreamIdentifier);
+
+    RefPtr<IPC::StreamConnectionWorkQueue> logWorkQueue() const { return m_logWorkQueue; }
```
shouldn't return a smart pointer

pvollan replied:
Thanks for catching! I will address this and further comments in the upcoming patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/NetworkProcess/LogStream.h:
```diff
@@ -0,0 +1,64 @@
+    void setup(uint64_t pid, IPC::StreamServerConnectionHandle&&, LogStreamIdentifier);
+
+    RefPtr<IPC::StreamConnectionWorkQueue> logWorkQueue() const { return m_logWorkQueue; }
+    RefPtr<IPC::StreamServerConnection> logStreamConnection() const { return m_logStreamConnection; }
```
ditto.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -832,65 +824,61 @@ static void registerLogHook()
 #ifdef NDEBUG
         // Don't send messages with types we don't want to log in release. Even though OS_LOG_TYPE_DEFAULT is the minimum,
         // the hook will be called for other subsystems with debug and info types.
-        if (type & (OS_LOG_TYPE_DEBUG | OS_LOG_TYPE_INFO))
```
So now we're sending info logging in release builds?

pvollan replied:
I will revert this in the next patch, to avoid extra logging.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -832,65 +824,61 @@ static void registerLogHook()
-        CString logCategory(msg->category);
-
-        auto qos = Thread::QOS::Background;
+        std::span logChannel(reinterpret_cast<const uint8_t*>(msg->subsystem), msg->subsystem ? strlen(msg->subsystem) + 1 : 0);
```
Can't we use byteCast?

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -832,65 +824,61 @@ static void registerLogHook()
-
-        auto qos = Thread::QOS::Background;
+        std::span logChannel(reinterpret_cast<const uint8_t*>(msg->subsystem), msg->subsystem ? strlen(msg->subsystem) + 1 : 0);
+        std::span logCategory(reinterpret_cast<const uint8_t*>(msg->category), msg->category ? strlen(msg->category) + 1 : 0);
```
ditto.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -832,65 +824,61 @@ static void registerLogHook()
-            os_log_message_s msg;
-            memset(&msg, 0, sizeof(msg));
+        char* messageString = os_log_copy_message_string(msg);
+        if (messageString) {
```
`if (char* messageString = os_log_copy_message_string(msg))`

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -832,65 +824,61 @@ static void registerLogHook()
-    WTFSignpostIndirectLoggingEnabled = true;
+void WebProcess::sendLogOnStream(std::span<const uint8_t>&& logChannel, std::span<const uint8_t>&& logCategory, std::span<uint8_t>&& logString, os_log_type_t type)
+{
+    if (!m_logStreamConnection)
```
Personally, I like this format in such cases:
```
if (RefPtr logStreamConnection = m_logStreamConnection)
    logStreamConnection->send(Messages::LogStream::LogOnBehalfOfWebContent(logChannel, logCategory, logString, type), m_logStreamIdentifier);
```

cdumez replied:
Also note elsewhere in this PR that you should have a RefPtr to the logStreamConnection *on the stack* before calling non-trivial functions on it, as per our smart pointer guidelines

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,97 @@
+
+LogStream::~LogStream()
+{
+    if (m_logStreamConnection) {
```
`if (RefPtr logStreamConnection = m_logStreamConnection) {`
then use `logStreamConnection` below, per smart pointer guidelines.

pvollan replied:
Good point! I will address this and the other comments in the next patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,97 @@
+        m_logStreamConnection->stopReceivingMessages(Messages::LogStream::messageReceiverName(), m_logStreamIdentifier->toUInt64());
+        m_logStreamConnection->invalidate();
+    }
+    if (m_logWorkQueue)
```
ditto, should probably be a RefPtr on the stack.

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,97 @@
+    // os_log_hook on sender side sends a null category and subsystem when logging to OS_LOG_DEFAULT.
+    auto osLog = OSObjectPtr<os_log_t>();
+    if (isNullTerminated(logSubsystem) && isNullTerminated(logCategory)) {
+        auto subsystem = reinterpret_cast<const char*>(logSubsystem.data());
```
byteCast?

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,97 @@
+    auto osLog = OSObjectPtr<os_log_t>();
+    if (isNullTerminated(logSubsystem) && isNullTerminated(logCategory)) {
+        auto subsystem = reinterpret_cast<const char*>(logSubsystem.data());
+        auto category = reinterpret_cast<const char*>(logCategory.data());
```
byteCast?

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    auto osLogPointer = osLog.get() ? osLog.get() : OS_LOG_DEFAULT;
+    auto logData = reinterpret_cast<const char*>(logString.data());
```
byteCast?

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,97 @@
+    m_logStreamIdentifier = logStreamIdentifier;
+    m_logStreamConnection = IPC::StreamServerConnection::tryCreate(WTFMove(serverConnection), { });
+    m_logWorkQueue = IPC::StreamConnectionWorkQueue::create("Log work queue"_s);
+    m_logStreamConnection->open(*m_logWorkQueue);
```
Need a smart pointer on the stack.

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,97 @@
+    m_logStreamConnection = IPC::StreamServerConnection::tryCreate(WTFMove(serverConnection), { });
+    m_logWorkQueue = IPC::StreamConnectionWorkQueue::create("Log work queue"_s);
+    m_logStreamConnection->open(*m_logWorkQueue);
+    m_logStreamConnection->startReceivingMessages(*this, Messages::LogStream::messageReceiverName(), m_logStreamIdentifier->toUInt64());
```
ditto.

cdumez commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -661,6 +666,12 @@ class WebProcess final : public AuxiliaryProcess
+#if ENABLE(LOGD_BLOCKING_IN_WEBCONTENT)
+    void registerLogHook();
+    void setupLogStream();
+    void sendLogOnStream(std::span<const uint8_t>&& logChannel, std::span<const uint8_t>&& logCategory, std::span<uint8_t>&& logString, os_log_type_t);
```
We don't need to pass span by r-value reference.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -836,61 +828,57 @@ static void registerLogHook()
-            msg.privdata_sz = privdata.size();
+        if (char* messageString = os_log_copy_message_string(msg)) {
+            std::span logString(byteCast<uint8_t>(messageString), strlen(messageString) + 1);
+            WebProcess::singleton().sendLogOnStream(WTFMove(logChannel), WTFMove(logCategory), WTFMove(logString), type);
```
You don't need WTFMove() for span.

cdumez commented on Source/WebKit/NetworkProcess/LogStream.cpp:
```diff
@@ -0,0 +1,99 @@
+#include "StreamConnectionWorkQueue.h"
+#include "StreamServerConnection.h"
+
+#define MESSAGE_CHECK RELEASE_ASSERT
```
What's this? a message check is not supposed to be a RELEASE_ASSERT. We never want to kill the recipient on bad IPC from the sender.

pvollan replied:
Oh, good catch! Removed in latest patch.

Thanks for reviewing!

pvollan committed:
Use streaming IPC for log forwarding
rdar://133129911
https://bugs.webkit.org/show_bug.cgi?id=277592

Reviewed by Chris Dumez.

Use streaming IPC for log forwarding to reduce power usage. Sending data over streaming IPC is more
efficient than sending using Mach messages.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::didFinishLoading):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setupLogStream):
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent): Deleted.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::setPreferenceValue):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::didReceiveResponse):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureNetworkProcessConnection):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/LaunchServicesDatabaseManager.mm:
(WebKit::LaunchServicesDatabaseManager::handleEvent):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerLogHook):
(WebKit::WebProcess::setupLogStream):
(WebKit::WebProcess::sendLogOnStream):
(WebKit::logQueue): Deleted.
(WebKit::registerLogHook): Deleted.

Canonical link: https://commits.webkit.org/284209@main
